code/goonstation/modules/power/power_transmission_laser/transmission_laser.dm

```diff
@@ -0,0 +1,521 @@
+		var/old_range = range
+		find_blocker()
+		if(range > old_range) // Create new lasers if the new blocker is further away
+			setup_lasers()
+
+	if(length(laser_effects))
+		for(var/obj/effect/transmission_beam in laser_effects)
+			for(var/atom/beamed in get_turf(transmission_beam))
+				atom_beam_effect(beamed)
+
+	if(!blocker)
+		sell_power(output_level * WATT_TICK_TO_JOULE)
+
+	charge -= output_level
+
+//// Selling defines are here
+#define MINIMUM_BAR 1
+#define PROCESS_CAP (20 - MINIMUM_BAR)
+
+#define A1_CURVE 20
+
+/obj/machinery/power/transmission_laser/proc/sell_power(joules)
+	var/mega_joules = joules / (1 MW)
+
+	var/generated_cash = (2 * mega_joules * PROCESS_CAP) / ((2 * mega_joules) + (PROCESS_CAP * A1_CURVE))
+	generated_cash += (4 * mega_joules * MINIMUM_BAR) / (4 * mega_joules + MINIMUM_BAR)
+	generated_cash = round(generated_cash)
+	if(generated_cash < 0)
+		return
+
+	total_energy += joules
+	total_earnings += generated_cash
+	generated_cash += unsent_earnings
+	unsent_earnings = generated_cash
+
+	var/datum/money_account/engineering_bank_account = GLOB.station_money_database.get_account_by_department(DEPARTMENT_ENGINEERING)
+	var/datum/money_account/cargo_bank_account = GLOB.station_money_database.get_account_by_department(DEPARTMENT_SUPPLY)
+
+	var/medium_cut = generated_cash * 0.25
+	var/high_cut = generated_cash * 0.75
+
+	cargo_bank_account.deposit_credits(medium_cut, "Transmission Laser Payout")
+	unsent_earnings -= medium_cut
+
+	engineering_bank_account.deposit_credits(high_cut, "Transmission Laser Payout")
+	unsent_earnings -= high_cut
+
+#undef A1_CURVE
+#undef PROCESS_CAP
+#undef MINIMUM_BAR
+
+// Beam related procs
+
+/obj/machinery/power/transmission_laser/proc/setup_lasers()
+	var/turf/last_step = get_step(get_front_turf(), dir)
+	// Create new lasers from the starting point to either the blocker or the edge of the map
+	for(var/num = 1 to range + 1)
+		if(!(locate(/obj/effect/transmission_beam/) in last_step))
+			var/obj/effect/transmission_beam/new_beam = new(last_step, src)
+			new_beam.host = src
+			new_beam.dir = dir
+			laser_effects += new_beam
+
+		last_step = get_step(last_step, dir)
+
+/obj/machinery/power/transmission_laser/proc/destroy_lasers()
+	for(var/obj/effect/transmission_beam/listed_beam as anything in laser_effects)
+		laser_effects -= listed_beam
+		qdel(listed_beam)
+
+/obj/machinery/power/transmission_laser/proc/shorten_beam()
+	for(var/obj/effect/transmission_beam/listed_beam as anything in laser_effects)
+		if(get_dist(get_front_turf(), listed_beam.loc) > range)
+			laser_effects -= listed_beam
+			qdel(listed_beam)
+
+/// Affect the atom according to it's type and the ouput power of the laser.
+/obj/machinery/power/transmission_laser/proc/atom_beam_effect(atom/beam_target)
+	if(QDELETED(beam_target) || istype(beam_target, /obj/structure/window))
+		return
+	var/mw_power = (output_number * power_format_multi_output) / (1 MW)
+	if(isliving(beam_target))
```

Comment:
Other than effects is there anything else that needs excepting? If it's just a single type that's the issue I'd rather do a type check.
Also, jaunted wizards and demons don't get exploded. Rev does cause explosions though it doesn't get meated, so I'll fix that

Reply:
Oh also, holoparas do get hit by lasers.

Reply:
I don't know exactly, I'm just spitballing a few things that tend to be problems. This is *probably* fine as a type check if need be.

Reply:
I tried a type check for obj/effect before calling the proc, and checked the mob isn't a rev, since it seems to be the only problematic mob.
Seems to work alright, and the proc isn't very complicated so I'd rather keep it all within transmission_laser.dm 
If any further exceptions pop up I'll just make it an atom proc and override it where needed as you suggested.

Reply:
Okay so I did more testing and I'm going to move it to overrides as suggested. There are just too many types of damage for different turfs and objects.

Reply:
done